src/ServiceGetPositionFK.c

```diff
@@ -0,0 +1,148 @@
+
+// shorten the typename a little, locally
+typedef motoros2_interfaces__srv__GetPositionFK_Request GetPositionFK_Request;
+typedef motoros2_interfaces__srv__GetPositionFK_Response GetPositionFK_Response;
+
+void Ros_ServiceGetPositionFK_Initialize()
+{
+    MOTOROS2_MEM_TRACE_START(svc_get_position_fk_init);
+
+    rcl_ret_t ret = rclc_service_init_default(&g_serviceGetPositionFK, &g_microRosNodeInfo.node,
+        ROSIDL_GET_SRV_TYPE_SUPPORT(motoros2_interfaces, srv, GetPositionFK),
+        SERVICE_NAME_GET_POSITION_FK);
+    motoRosAssert_withMsg(ret == RCL_RET_OK, SUBCODE_FAIL_INIT_SERVICE_GET_POSITION_FK,
+        "Failed to init service (%d)", (int)ret);
+
+    rosidl_runtime_c__String__init(&g_messages_GetPositionFK.response.message);
+    geometry_msgs__msg__PoseStamped__init(&g_messages_GetPositionFK.response.result);
+
+    MOTOROS2_MEM_TRACE_REPORT(svc_get_position_fk_init);
+}
+
+void Ros_ServiceGetPositionFK_Cleanup()
+{
+    MOTOROS2_MEM_TRACE_START(svc_get_position_fk_fini);
+
+    rcl_ret_t ret;
+
+    Ros_Debug_BroadcastMsg("Cleanup service " SERVICE_NAME_GET_POSITION_FK);
+    ret = rcl_service_fini(&g_serviceGetPositionFK, &g_microRosNodeInfo.node);
+    if (ret != RCL_RET_OK)
+        Ros_Debug_BroadcastMsg(
+            "Failed cleaning up " SERVICE_NAME_GET_POSITION_FK " service: %d", ret);
+    rosidl_runtime_c__String__fini(&g_messages_GetPositionFK.response.message);
+    geometry_msgs__msg__PoseStamped__fini(&g_messages_GetPositionFK.response.result);
+
+    MOTOROS2_MEM_TRACE_REPORT(svc_get_position_fk_fini);
+}
+
+//TODO(gavanderhoorn): refactor: create M+ Cart types <-> ROS types conversion lib
+static void MpCoord_To_GeomMsgsPose(MP_COORD const* const mp_coord, geometry_msgs__msg__Pose* const geom_pose)
+{
+    geom_pose->position.x = NANOMETERS_TO_METERS(mp_coord->x);
+    geom_pose->position.y = NANOMETERS_TO_METERS(mp_coord->y);
+    geom_pose->position.z = NANOMETERS_TO_METERS(mp_coord->z);
+    QuatConversion_MpCoordOrient_To_GeomMsgsQuaternion(
+        mp_coord->rx, mp_coord->ry, mp_coord->rz,
+        &geom_pose->orientation);
+}
+
+void Ros_ServiceGetPositionFK_Trigger(const void* request_msg, void* response_msg)
+{
+    GetPositionFK_Request* request = (GetPositionFK_Request*) request_msg;
+    GetPositionFK_Response* response = (GetPositionFK_Response*) response_msg;
+
+    Ros_Debug_BroadcastMsg("%s: entry", __func__);
+
+    //init response message
+    response->success = FALSE;
+    response->result_code = 0;
+    rosidl_runtime_c__String__assign(&response->message, "");
+    rosidl_runtime_c__String__assign(
+        &g_messages_GetPositionFK.response.result.header.frame_id, "");
+
+    Ros_Debug_BroadcastMsg("%s: requested: grp no: %u, tool no: %u",
+        __func__, request->group_no, request->tool_no);
+
+    //make sure a valid group is specified (group_no is unsigned, so cannot be < 0)
+    if (request->group_no >= g_Ros_Controller.numGroup)
```

Comment:
just realised this doesn't work for multi-group systems with integrated axes (ie: robot on track). Unless those combined groups also get a group nr.